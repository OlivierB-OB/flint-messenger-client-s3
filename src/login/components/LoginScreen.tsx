import Container from '@material-ui/core/Container';
import Tab from '@material-ui/core/Tab';
import Tabs from '@material-ui/core/Tabs';
import React, { useState, useEffect } from 'react';
import { connect } from 'react-redux';
import { Action } from 'redux';
import { RegistrationForm } from '../../profileForm/components/RegistrationForm';
import { IProfileFormStatus } from '../../profileForm/types';
import { Loading } from '../../layout/components/Loading';
import { IAppState } from '../../appReducer';
import { ILoginStatus } from '../types';
import { LoginTabPanel } from './LoginTabPanel';
import { LoginForm } from './LoginForm';
import { resetProfileFormContent } from '../../profileForm/actions/resetProfileFormContent';
import { makeExitApplication } from '../../layout/actions/makeExitApplication';
import { ThunkDispatch } from 'redux-thunk';

export interface ILoginScreenDisplayProps {
  loginStatus: ILoginStatus;
  registrationStatus: IProfileFormStatus;
  resetRegistrationForm(): void;
}

export function LoginScreenDisplay({
  loginStatus,
  registrationStatus,
  resetRegistrationForm,
}: ILoginScreenDisplayProps) {
  const [tab, setTab] = useState(0);
  useEffect(() => void resetRegistrationForm(), [resetRegistrationForm]);
  if ([loginStatus, registrationStatus].includes('unavailable')) return <Loading />;
  return (
    <Container maxWidth="lg" style={{ marginTop: '2rem' }}>
      <Tabs
        value={tab}
        onChange={(_, newTab) => setTab(newTab)}
        indicatorColor="primary"
        textColor="primary"
        variant="fullWidth"
      >
        <Tab label="Login" />
        <Tab label="Register" />
      </Tabs>
      <LoginTabPanel value={tab} index={0}>
        <LoginForm />
      </LoginTabPanel>
      <LoginTabPanel value={tab} index={1}>
        <RegistrationForm />
      </LoginTabPanel>
    </Container>
  );
}

const mapStateToProps = ({ login, profileForm }: IAppState) => ({
  loginStatus: login.status,
  registrationStatus: profileForm.status,
});

const mapDispatchToProps = (dispatch: ThunkDispatch<IAppState, void, Action>) => ({
  exitApplication: () => dispatch(makeExitApplication()),
  resetRegistrationForm: () => dispatch(resetProfileFormContent()),
});

export const LoginScreen = connect(mapStateToProps, mapDispatchToProps)(LoginScreenDisplay);
