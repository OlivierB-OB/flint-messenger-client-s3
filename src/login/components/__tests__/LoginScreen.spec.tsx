import React from 'react';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { shallow } from 'enzyme';
import { LoginScreen } from '../LoginScreen';

const middlewares = [thunk];
const mockStore = configureMockStore(middlewares);

describe('LoginScreen', () => {
  let Connected: any = LoginScreen;

  it('should provide login status, registration status, callbacks to LoginScreenDisplay', async () => {
    const store = mockStore({
      login: {
        status: 'login status',
      },
      profileForm : {
        status: 'login status',
      },
    });
    const component = shallow(<Connected store={store} />).dive();
    expect(component).toMatchSnapshot();

    (component.prop('resetRegistrationForm') as any)();
    expect(store.getActions()).toEqual([
      {
        info: undefined,
        type: 'RESET_PROFILE_FORM_CONTENT',
      },
    ]);
    store.clearActions();
  });
});
