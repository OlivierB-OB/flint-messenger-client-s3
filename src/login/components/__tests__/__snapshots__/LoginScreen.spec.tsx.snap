// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginScreen should provide login status, registration status, callbacks to LoginScreenDisplay 1`] = `
<LoginScreenDisplay
  exitApplication={[Function]}
  loginStatus="login status"
  registrationStatus="login status"
  resetRegistrationForm={[Function]}
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
/>
`;
